File: plugins/logger.go
package plugins

import (
	"context"
	"fmt"
	"os"
	"sync"
	"time"
)

const (
	LogLevelQuiet        = 0 // errors and doc-change info only
	LogLevelPluginDetail = 1 // short-summary — less than summary
	LogLevelSummary      = 2 // + pipeline phase timing totals
	LogLevelVerbose      = 3 // + per-plugin spawn and per-hook timing
)

type Logger struct {
	level  int
	mu     sync.Mutex
	timers map[string]time.Time
}

func NewLogger(level int) *Logger {
	return &Logger{
		level:  level,
		timers: make(map[string]time.Time),
	}
}

// parseLogLevel converts a DB log level string (e.g. "SUMMARY") to its integer constant.
func parseLogLevel(s string) int {
	switch s {
	case "QUIET":
		return LogLevelQuiet
	case "SUMMARY":
		return LogLevelSummary
	case "SHORT_SUMMARY":
		return LogLevelPluginDetail
	case "FULL":
		return LogLevelVerbose
	default:
		return LogLevelSummary
	}
}

// Logger returns a logger seeded from the project's configured log level.
func (db DatabasePool[PluginConfig]) Logger(ctx context.Context) (*Logger, error) {
	config, err := db.ProjectConfig(ctx)
	if err != nil {
		return nil, err
	}
	return NewLogger(parseLogLevel(config.LogLevel)), nil
}

// Error always prints to stderr regardless of log level.
func (l *Logger) Error(format string, args ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", args...)
}

// Warn always prints to stderr regardless of log level.
func (l *Logger) Warn(format string, args ...any) {
	fmt.Fprintf(os.Stderr, format+"\n", args...)
}

// Info prints to stderr only when the logger's level is >= minLevel.
func (l *Logger) Info(minLevel int, format string, args ...any) {
	if l.level >= minLevel {
		fmt.Fprintf(os.Stderr, format+"\n", args...)
	}
}

// Time starts a named timer.
func (l *Logger) Time(label string) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.timers[label] = time.Now()
}

// TimeEnd stops the named timer and prints the elapsed time if level >= minLevel.
func (l *Logger) TimeEnd(label string, minLevel int) {
	l.mu.Lock()
	start, ok := l.timers[label]
	if ok {
		delete(l.timers, label)
	}
	l.mu.Unlock()

	if ok && l.level >= minLevel {
		ms := float64(time.Since(start).Nanoseconds()) / 1e6
		fmt.Fprintf(os.Stderr, "  %s: %.1fms\n", label, ms)
	}
}

// At reports whether the logger's level is at or above minLevel.
func (l *Logger) At(minLevel int) bool {
	return l.level >= minLevel
}

File: packages/houdini/src/lib/logger.ts
import { LogLevel } from './types.js'

const LEVEL_ORDER: Record<string, number> = {
	quiet: 0,
	'short-summary': 1,
	summary: 2,
	full: 3,
}

export class Logger {
	private timers = new Map<string, number>()
	private order: number

	constructor(public readonly level: string = LogLevel.Summary) {
		this.order = LEVEL_ORDER[level] ?? 1
	}

	error(msg: string, ...args: unknown[]) {
		console.error(msg, ...args)
	}

	warn(msg: string, ...args: unknown[]) {
		console.warn(msg, ...args)
	}

	info(msg: string, minLevel: string = LogLevel.Summary) {
		if (this.order >= (LEVEL_ORDER[minLevel] ?? 1)) {
			console.log(msg)
		}
	}

	time(label: string) {
		this.timers.set(label, performance.now())
	}

	timeEnd(label: string, minLevel: string) {
		const start = this.timers.get(label)
		if (start === undefined) return
		this.timers.delete(label)
		if (this.order >= (LEVEL_ORDER[minLevel] ?? 1)) {
			const ms = (performance.now() - start).toFixed(1)
			console.log(`  ${label}: ${ms}ms`)
		}
	}

	at(minLevel: string): boolean {
		return this.order >= (LEVEL_ORDER[minLevel] ?? 1)
	}
}
